Fix -Zdump-dep-graph

Mark-Simulacrum opened the pull request:
This modifies the query "modifier" macros to support additional modifiers and introduces a new one dedicated to identifying queries with parameters. It also explicitly drops considering queries with `()` parameters to be "has_params" -- this feels like what the code *wanted* but I would like a second pair of eyes on it. I believe all queries that are declared via the macro infrastructure were previously has_params = true; the previous macro definition always would have found parameters as far as I can tell.

This is all with the goal of fixing -Zdump-dep-graph, which seems to have broken a while ago (I didn't try to bisect). Currently even on a `fn main() {}` file it'll ICE (you need to also pass -Zquery-dep-graph for it to work at all), and this patch indirectly fixes the cause of that ICE. Also add a test for that trivial condition of "produces files"; don't try to inspect those outputs though.

rust-highfive commented:
r? @matthewjasper

(rust-highfive has picked a reviewer for you, use r? to override)

Mark-Simulacrum commented:
cc @rust-lang/wg-incr-comp -- I'm not sure if folks recall the intent of has_params, which might be helpful to determine if this is the "right" thing to do.

Mark-Simulacrum commented on compiler/rustc_middle/src/ty/query.rs:
```diff
@@ -114,8 +114,8 @@ macro_rules! query_storage {
     };
-    ([$other:ident $(($($other_args:tt)*))* $(, $($modifiers:tt)*)*][$($args:tt)*]) => {
-        query_storage!([$($($modifiers)*)*][$($args)*])
+    ([$modifier:tt $($rest:tt)*][$($args:tt)*]) => {
```
FWIW as noted in the commit message, I don't really fully understand what is going on here - I was getting pretty weird errors without the first commit though. You can see a pretty minimal example of this here - https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=de08d72c9a23062bf3d224f0d873d27d.

I think the changes to the macro are correct, but I'm not an expert with macro_rules so I definitely could have gotten this wrong. Happy to adjust!

cjgillot replied:
This is a known bug in the macro. 0358edeb546dbdc9d2f894c19b54c07884d9b9c8 should fix it.

Mark-Simulacrum replied:
Hm, the first commit in this PR looks simpler than that one, but happy to rebase over that (not sure where that's from -- GitHub isn't showing a PR link).

cjgillot replied:
It's from #85830.

cjgillot commented:
IIUC, the crash happens when trying to debug-print a DepNode whch parameter () by mistakenly trying to extract a DefId from a Fingerprint which is not a proper DefPathHash. The bug is therefore in `rustc_middle::dep_graph::dep_node::DepNodeExt::extract_def_id` which tests `can_reconstruct_query_key()` to know whether the query key is a `CrateNum`/`LocalDefId`/`DefId`. I think this PR hides the problem rather than solving it. The correct fix would be to create a separate `has_def_path_hash` field in `DepKindStruct` which tests whether the query key is any of the three, and equivalently if a DefId can be extracted from the fingerprint.

Mark-Simulacrum commented:
Hm, that seems like a good thing to tackle, I agree. I think this PR is still desirable (or we should just remove has_params), since it seems like the meaning of that is not too clear right now.

I'll have to dig into the 2 manually declared queries with has_params: false, and see if that's significant for them...

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3777318677) [(plain)](https://github.com/rust-lang/rust/commit/3d73c4d3d5367518aecc03d9c57a6c807066f44a/checks/3777318677/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling rustc_session v0.0.0 (/checkout/compiler/rustc_session)
   Compiling rustc_hir_pretty v0.0.0 (/checkout/compiler/rustc_hir_pretty)
   Compiling rustc_attr v0.0.0 (/checkout/compiler/rustc_attr)
   Compiling rustc_query_system v0.0.0 (/checkout/compiler/rustc_query_system)
error[E0600]: cannot apply unary operator `!` to type `FingerprintStyle`
  --> compiler/rustc_query_system/src/dep_graph/dep_node.rs:78:16
   |
78 |             if !kind.fingerprint_style()
   |                ^^^^^^^^^^^^^^^^^^^^^^^^^ cannot apply unary operator `!`
   |
   = note: an implementation of `std::ops::Not` might be missing for `FingerprintStyle`
For more information about this error, try `rustc --explain E0600`.
error: could not compile `rustc_query_system` due to previous error
warning: build failed, waiting for other jobs to finish...
error: build failed
```

</details>

Mark-Simulacrum commented:
Alright, replaced pretty much the whole PR with different patch -- and extracted https://github.com/rust-lang/rust/pull/89466 for the base macro changes which are not actually needed for this PR now.

I looked into fully removing has_params but didn't end up making that change. The only uses left are the debug assert in DepKind::new_no_params and the debug_node method. The latter could be removed with no loss, I believe -- just a difference in how nodes look (printed as `foo` today, `foo()` with that condition dropped). Since has_params is true currently for all but Null and TraitSelect DepKinds (neither of which is a proper query), it's probably the case that this could be refactored. But seems like future work.

cjgillot commented on compiler/rustc_middle/src/dep_graph/dep_node.rs:
```diff
@@ -350,14 +352,16 @@ impl DepNodeExt for DepNode {
+        match kind.fingerprint_style() {
+            FingerprintStyle::Opaque => Err(()),
+            FingerprintStyle::Unit => {
+                if !kind.has_params {
```
`FingerprintStyle::Unit` should now be equivalent to `has_params == false`, isn't it? We should be able to get rid of `has_params`.

Mark-Simulacrum replied:
Not currently. A query with `()` key currently "has_params"; the only two things to not have params are Null and TraitSelect which are not generated through the normal macro infrastructure. dropped the "fix" for that from this branch for now, though I can re-add it, it's not big (b2b53bb9a4f2f5e1f96f8f01ad7b378f5eb1b4d1).

cjgillot commented:
@bors r+

bors commented:
:pushpin: Commit 63aaf882c361feb26cf8fb21a3ef2f83c2d1ef51 has been approved by `cjgillot`

<!-- @bors r=cjgillot 63aaf882c361feb26cf8fb21a3ef2f83c2d1ef51 -->
<!-- homu: {"type":"Approved","sha":"63aaf882c361feb26cf8fb21a3ef2f83c2d1ef51","approver":"cjgillot"} -->

workingjubilee commented:
This looks like yours? Apologies if I am wrong.
```
 ------------------------------------------
RUST_DEP_GRAPH=/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/dep-graph/dep-graph/dep-graph LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/dep-graph/dep-graph:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/dep-graph/dep-graph -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/dep-graph/dep-graph  -Clinker='arm-none-eabi-gcc' \
        -Cincremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/dep-graph/dep-graph/incr \
        -Zquery-dep-graph -Zdump-dep-graph foo.rs
------------------------------------------
stderr:
------------------------------------------
------------------------------------------
error: linking with `arm-none-eabi-gcc` failed: exit status: 1
  |
  = note: "arm-none-eabi-gcc" "-m64" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/dep-graph/dep-graph/foo.18u5b2jmv51dofh3.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/dep-graph/dep-graph/foo.25sg5ox0q16b7auf.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/dep-graph/dep-graph/foo.3c6c78fzwfnp5rr9.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/dep-graph/dep-graph/foo.3pwxpcv7ig8jfb30.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/dep-graph/dep-graph/foo.3wccqld67w709v08.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/dep-graph/dep-graph/foo.4rlnbcoxr7ik1e3w.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/dep-graph/dep-graph/foo.4unz5qo9qikgp26p.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/dep-graph/dep-graph/foo.50rnet0thytpgybx.rcgu.o" "-Wl,--as-needed" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/dep-graph/dep-graph" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,--start-group" "-Wl,-Bstatic" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-5548ef64e6092712.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-a8f484438681c15e.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libminiz_oxide-b69fd8507c8409af.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libadler-2dff396b99681a6b.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libobject-a78bd5aa183c7115.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libmemchr-846c64d5a2fbc5ee.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libaddr2line-24c09062b6dc787b.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgimli-ce93f351ca41a57b.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_detect-5b404d65540eadf8.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-514cb174319eb6d4.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libhashbrown-37c1534c42bc4f8c.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-06d0f7780fda2fd9.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-126d4eb258557375.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-88bb1ba4dd271224.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-b4a707e310d7480c.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-0b5a376b429677a6.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-2a6a2797f7a73818.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-0e3656b1fda5fd7b.rlib" "-Wl,--end-group" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-eecefd843a0dbc02.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/dep-graph/dep-graph/foo" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro" "-Wl,-znow" "-nodefaultlibs"
  = note: arm-none-eabi-gcc: error: unrecognized command line option '-m64'
```
@bors r-

bors commented:
:umbrella: The latest upstream changes (presumably #89266) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:pushpin: Commit 85c4fd81799de01e8eb3a84f96a13baafc8de407 has been approved by `cjgillot`

<!-- @bors r=cjgillot 85c4fd81799de01e8eb3a84f96a13baafc8de407 -->
<!-- homu: {"type":"Approved","sha":"85c4fd81799de01e8eb3a84f96a13baafc8de407","approver":"cjgillot"} -->

rust-log-analyzer commented:
The job **`dist-various-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/3845965222) [(plain)](https://github.com/rust-lang-ci/rust/commit/8f2565b630f71928671276797f4659fcc679d795/checks/3845965222/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test [run-make] run-make/rustc-macro-dep-files ... ok
test [run-make] run-make/env-dep-info ... ok
test [run-make] run-make/emit-shared-files ... ok
test [run-make] run-make/thumb-none-cortex-m ... ok
Some tests failed in compiletest suite=run-make mode=run-make host=x86_64-unknown-linux-gnu target=thumbv6m-none-eabi

failures:


---- [run-make] run-make/dep-graph stdout ----
error: make failed
status: exit status: 2
command: "make"
stdout:
stdout:
------------------------------------------
RUST_DEP_GRAPH=/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/dep-graph/dep-graph/dep-graph LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/dep-graph/dep-graph:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/dep-graph/dep-graph -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/dep-graph/dep-graph  -Clinker='arm-none-eabi-gcc' \
        -Cincremental=/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/dep-graph/dep-graph/incr \
        --target thumbv6m-none-eabi \
        -Zquery-dep-graph -Zdump-dep-graph foo.rs
------------------------------------------
stderr:
------------------------------------------
error[E0463]: can't find crate for `std`
error[E0463]: can't find crate for `std`
  |
  = note: the `thumbv6m-none-eabi` target may not support the standard library
  = note: `std` is required by `<unknown>` because it does not declare `#![no_std]`
  = help: consider building the standard library from source with `cargo build -Zbuild-std`
error: aborting due to previous error

For more information about this error, try `rustc --explain E0463`.
For more information about this error, try `rustc --explain E0463`.
make: *** [Makefile:6: all] Error 1
------------------------------------------




failures:
    [run-make] run-make/dep-graph

test result: FAILED. 14 passed; 1 failed; 24 ignored; 0 measured; 0 filtered out; finished in 21.41s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/thumbv6m-none-eabi/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--rustdoc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "--src-base" "/checkout/src/test/run-make" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make" "--stage-id" "stage2-thumbv6m-none-eabi" "--suite" "run-make" "--mode" "run-make" "--target" "thumbv6m-none-eabi" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/bin/FileCheck" "--linker" "arm-none-eabi-gcc" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/thumbv6m-none-eabi/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--llvm-version" "13.0.0-rust-1.57.0-nightly" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker dwp engine executionengine extensions filecheck frontendopenacc frontendopenmp fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irreader jitlink libdriver lineeditor linker lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit orcshared orctargetprocess passes powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info xray" "--llvm-bin-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/bin" "--cc" "arm-none-eabi-gcc" "--cxx" "arm-none-eabi-g++" "--cflags" "-ffunction-sections -fdata-sections -mthumb -march=armv6s-m" "--ar" "arm-none-eabi-ar" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:15:12
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3846022039)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/3846022039","builder_name":"checks-actions"} -->

Mark-Simulacrum commented:
@bors r=cjgillot rollup=iffy

bors commented:
:pushpin: Commit 415a9a2ea69e917c5af69a484a602010ccf67198 has been approved by `cjgillot`

<!-- @bors r=cjgillot 415a9a2ea69e917c5af69a484a602010ccf67198 -->
<!-- homu: {"type":"Approved","sha":"415a9a2ea69e917c5af69a484a602010ccf67198","approver":"cjgillot"} -->

bors commented:
:hourglass: Testing commit 415a9a2ea69e917c5af69a484a602010ccf67198 with merge 15491d7b6be1065217e09bd735e4ecffbe9838ba...
<!-- homu: {"type":"BuildStarted","head_sha":"415a9a2ea69e917c5af69a484a602010ccf67198","merge_sha":"15491d7b6be1065217e09bd735e4ecffbe9838ba"} -->

rust-timer commented:
Finished benchmarking commit (15491d7b6be1065217e09bd735e4ecffbe9838ba): [comparison url](https://perf.rust-lang.org/compare.html?start=bb918d0a5bf22211df0423f7474e4e4056978007&end=15491d7b6be1065217e09bd735e4ecffbe9838ba).

**Summary**: This benchmark run did not return any relevant changes.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression
